### PR TITLE
If the zipfile is corrupt it should remove out from cache

### PR DESCRIPTION
**Pull request by Forestsoft-de**
Its not possible to fix it if you dont know that there is a cache. Corrupt zipfiles shouldnt be cached.

Btw: Unittests for complete LocalCache class. The Plugininstaller private methods makes it impossibe to unit test.
The verification that cacheclear is called in case of exception is missed.

### CODE REVIEW

File: src/Shyim/PluginInstaller.php [hunk 1]
```diff
@@ -229,18 +258,20 @@ private static function downloadAndMovePlugin($url, $name, $version)
...
+            $zip->open($zipFile);
+            $folderpath = str_replace("\\", "/", $zip->statIndex(0)['name']);
+            $pos = strpos($folderpath, '/');
+            $path = substr($folderpath, 0, $pos);
+            $location = self::getExtractLocation($path);
+            $zip->extractTo($location);
+            $zip->close();
+        } catch (\Exception $e) {
+            LocalCache::cleanByPath($zipFile);
+            throw new \Exception(sprintf("Could not extract Plugin %s", $zipFile));
```
Comment (shyim) [inline src/Shyim/PluginInstaller.php]: Maybe print exception message

File: tests/LocalCacheTest.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    private $root;
+
+    protected function setUp()
+    {
+        parent::setUp();
+        $this->root = vfsStream::setUp("cache");
+    }
+
+
+    /**
```
Comment (shyim) [inline tests/LocalCacheTest.php]: Can you remove these empty phpdocs?

File: src/Shyim/PluginInstaller.php [hunk 3]
```diff
@@ -208,14 +230,21 @@ private static function apiRequest($path, $method, array $params = [])
...
     {
-        $ch = curl_init($url);
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
-        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
-        $content = curl_exec($ch);
-        curl_close($ch);
+        $content = self::makePluginHTTPRequest($url);
+
+        if (substr($content, 0, 1) == "{" && substr($content, strlen($content) - 1, 1) == "}") {
+            $jsonMessage = json_decode($content, true);
```
Comment (shyim) [inline src/Shyim/PluginInstaller.php]: You have now two things in this pull request

File: src/Shyim/PluginInstaller.php (hunk 1 above)
Comment (Forestsoft-de) [reply src/Shyim/PluginInstaller.php]: now i use your exception handler :+1:

File: tests/LocalCacheTest.php (hunk 2 above)
Comment (Forestsoft-de) [reply tests/LocalCacheTest.php]: sure see latest pr

File: tests/LocalCacheTest.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
+<?php
+/**
```
Comment (shyim) [inline tests/LocalCacheTest.php]: Can you also remove the PhpStorm default header?

**Comment by shyim** [comment]
Can you also rebase, i have merged your another PR

File: tests/LocalCacheTest.php (hunk 4 above)
Comment (Forestsoft-de) [reply tests/LocalCacheTest.php]: removed

**Review by Forestsoft-de — COMMENTED** [review]
what should i rebase? the multiple commits or ur master?

File: src/Shyim/PluginInstaller.php [hunk 5]
```diff
@@ -208,14 +230,21 @@ private static function apiRequest($path, $method, array $params = [])
...
     private static function downloadAndMovePlugin($url, $name, $version)
     {
-        $ch = curl_init($url);
-        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
-        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
-        $content = curl_exec($ch);
-        curl_close($ch);
+        $content = self::makePluginHTTPRequest($url);
+
+        if (substr($content, 0, 1) == "{" && substr($content, strlen($content) - 1, 1) == "}") {
```
Comment (shyim) [inline src/Shyim/PluginInstaller.php]: This part should be not in this Pull Request

**Review by shyim — CHANGES_REQUESTED** [review]

File: src/Shyim/PluginInstaller.php (hunk 5 above)
Comment (Forestsoft-de) [reply src/Shyim/PluginInstaller.php]: just rebased my master. should now clean :)

**Comment by shyim** [comment]
Thanks :)

**Pull request merged**